File: sdk/arena-python-sdk/arenasdk/training/job_builder.py
#!/usr/bin/env python
from __future__ import annotations
import abc 
from arenasdk.exceptions.arena_exception import ArenaException
from arenasdk.fields.fields import *
from arenasdk.enums.types import ArenaErrorType
from arenasdk.enums.types import TrainingJobType
from arenasdk.training.job import TrainingJob
from typing import List
from typing import Dict

def process_items(items: List[str]):
    for item in items:
        print(item) 

class JobBuilder(metaclass=abc.ABCMeta):
    def __init__(self,job_type: TrainingJobType):
        self._job_type: TrainingJobType = job_type
        self._job_name = ""
        self._options: List[Field] = list()
        self._command = ""
        
    def build(self) -> TrainingJob:
        args = list()
        try:
            for field in self._options:
                if not isinstance(field,Field):
                    raise ArenaException(ArenaErrorType.Unknown,"the type of option {} is not Field".format(field))
                field.validate()
                for opt in field.options():
                    args.append(opt)
            return TrainingJob(self._job_name,self._job_type,args,self._command)
        except ArenaException as e:
            raise e 
    
    
    def with_name(self,name: str) -> JobBuilder:
        self._job_name = name
        self._options.append(StringField("--name",name))
        return self 

    def with_image(self,image: str) -> JobBuilder:
        self._options.append(StringField("--image",image))
        return self 

    def with_workers(self,count: int) -> JobBuilder:
        self._options.append(StringField("--workers",str(count)))
        return self

    def with_image_pull_secrets(self,secrets: List[str]) -> JobBuilder:
        self._options.append(StringListField("--image-pull-secret",secrets))
        return self
    
    def with_gpus(self,count: int) -> JobBuilder:
        self._options.append(StringField("--gpus",str(count)))
        return self

    def with_envs(self,envs: Dict[str,str]) -> JobBuilder:
        self._options.append(StringMapField("--env",envs,"="))
        return self

    def with_node_selectors(self,selectors: Dict[str, str]) -> JobBuilder:
        self._options.append(StringMapField("--selector",selectors,"="))
        return self

    def with_tolerations(self,tolerations: List[str]) -> JobBuilder:
        self._options.append(StringListField("--toleration",tolerations))
        return self
    
    def with_config_files(self,files: Dict[str, str]) -> JobBuilder:
        self._options.append(StringMapField("--config-file",files,":"))
        return self 

    def with_annotations(self,annotions: Dict[str, str]) -> JobBuilder:
        self._options.append(StringMapField("--annotation",annotions,"="))
        return self

    def with_datas(self,datas: Dict[str,str]) -> JobBuilder:
        self._options.append(StringMapField("--data",datas,":"))
        return self 
    
    def with_data_dirs(self,data_dirs: Dict[str, str]) -> JobBuilder:
        self._options.append(StringMapField("--data-dir",data_dirs,":"))
        return  self

    def with_log_dir(self,dir: str) -> JobBuilder:
        self._options.append(StringField("--logdir",dir))
        return self

    def with_priority(self,priority: str) -> JobBuilder:
        self._options.append(StringField("--priority",priority))
        return  self
    
    def enable_rdma(self) -> JobBuilder:
        self._options.append(BoolField("--rdma"))
        return  self
    
    def with_sync_image(self,image: str) -> JobBuilder:
        self._options.append(StringField("--sync-image",image))
        return  self

    def with_sync_mode(self,mode: str) -> JobBuilder:
        self._options.append(StringField("--sync-mode",mode))
        return  self 

    def with_sync_source(self,source: str) -> JobBuilder:
        self._options.append(StringField("--sync-source",source))
        return  self
    
    def enable_tensorboard(self) -> JobBuilder:
        self._options.append(BoolField("--tensorboard"))
        return self 

    def with_tensorboard_image(self,image: str) -> JobBuilder:
        self._options.append(StringField("--tensorboard-image",image))
        return self

    def with_working_dir(self,dir: str) -> JobBuilder:
        self._options.append(StringField("--working-dir",dir))
        return self

    def with_retry_count(self,count: int) -> JobBuilder:
        self._options.append(StringField("--retry",str(count)))
        return self 

    def enable_coscheduling(self) -> JobBuilder:
       self._options.append(BoolField("--gang"))
       return self

    def with_command(self,command: str) -> JobBuilder:
        self._command = command
        return self

File: sdk/arena-java-sdk/src/main/java/com/github/kubeflow/arena/model/training/JobBuilder.java
package com.github.kubeflow.arena.model.training;

import com.github.kubeflow.arena.enums.TrainingJobType;
import com.github.kubeflow.arena.exceptions.ArenaException;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import com.github.kubeflow.arena.model.fields.*;

public abstract class JobBuilder {

    protected String jobName;

    protected TrainingJobType jobType;

    protected ArrayList<Field> options;

    protected String command;

    public JobBuilder(TrainingJobType jobType){
        this.jobType = jobType;
        this.options =  new ArrayList<Field>();
    }

    public TrainingJob build() throws ArenaException {
        List<String> args = new ArrayList<>();
        for (int i = 0;i < this.options.size();i++) {
            Field f = this.options.get(i);
            f.validate();
            for(int j = 0;j < f.options().size();j++) {
                args.add(f.options().get(j));
            }
        }
        return new TrainingJob(this.jobName,this.jobType,args,this.command);
    }

    public JobBuilder name(String name) {
        this.jobName = name;
        this.options.add(new StringField("--name",name));
        return this;
    }

    public JobBuilder image(String image) {
        this.options.add(new StringField("--image",image));
        return this;
    }

    public JobBuilder workers(int count) {
        this.options.add(new StringField("--workers",String.valueOf(count)));
        return this;
    }

    public JobBuilder imagePullSecrets(ArrayList<String> secrets) {
        this.options.add(new StringListField("--image-pull-secret",secrets));
        return this;
    }

    public JobBuilder imagePullPolicy(String policy) {
        this.options.add(new StringField("--image-pull-policy", policy));
        return this;
    }

    public JobBuilder gpus(int count) {
        this.options.add(new StringField("--gpus",String.valueOf(count)));
        return this;
    }

    public JobBuilder envs(Map<String, String> envs) {
        this.options.add(new StringMapField("--env",envs,"="));
        return this;
    }

    public JobBuilder nodeSelectors(Map<String, String> selectors) {
        this.options.add(new StringMapField("--selector",selectors,"="));
        return this;
    }

    public JobBuilder tolerations(ArrayList<String> tolerations) {
        this.options.add(new StringListField("--toleration",tolerations));
        return this;
    }

    public JobBuilder configFiles(Map<String, String> files) {
        this.options.add(new StringMapField("--config-file",files,":"));
        return this;
    }

    public JobBuilder labels(Map<String, String> labels) {
        this.options.add(new StringMapField("--label", labels,"="));
        return this;
    }

    public JobBuilder devices(Map<String, String> devices) {
        this.options.add(new StringMapField("--device", devices,"="));
        return this;
    }

    public JobBuilder annotations(Map<String, String> annotations) {
        this.options.add(new StringMapField("--annotation",annotations,"="));
        return this;
    }

    public JobBuilder datas(Map<String, String> datas) {
        this.options.add(new StringMapField("--data",datas,":"));
        return this;
    }

    public JobBuilder dataDirs(Map<String, String> dataDirs) {
        this.options.add(new StringMapField("--data-dir",dataDirs,":"));
        return  this;
    }

    public JobBuilder logDir(String dir) {
        this.options.add(new StringField("--logdir",dir));
        return this;
    }

    public JobBuilder priority(String priority) {
        this.options.add(new StringField("--priority",priority));
        return  this;
    }

    public JobBuilder enableRDMA() {
        this.options.add(new BoolField("--rdma"));
        return  this;
    }

    public JobBuilder syncImage(String image) {
        this.options.add(new StringField("--sync-image",image));
        return  this;
    }

    public JobBuilder syncMode(String mode) {
        this.options.add(new StringField("--sync-mode",mode));
        return  this;
    }

    public JobBuilder syncSource(String source) {
        this.options.add(new StringField("--sync-source",source));
        return  this;
    }

    public JobBuilder enableTensorboard() {
        this.options.add(new BoolField("--tensorboard"));
        return this;
    }

    public JobBuilder tensorboardImage(String image) {
        this.options.add(new StringField("--tensorboard-image",image));
        return this;
    }

    public JobBuilder workingDir(String dir) {
        this.options.add(new StringField("--working-dir",dir));
        return this;
    }

    public JobBuilder retryCount(int count) {
        this.options.add(new StringField("--retry",String.valueOf(count)));
        return this;
    }

    public JobBuilder enableCoscheduling() {
       this.options.add(new BoolField("--gang"));
       return this;
    }

    public JobBuilder shell(String shell) {
        this.options.add(new StringField("--shell",String.valueOf(shell)));
        return this;
    }

    public JobBuilder command(String command) {
        this.command = command;
        return this;
    }
}


